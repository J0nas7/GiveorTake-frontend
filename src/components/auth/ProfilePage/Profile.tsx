"use client";

// External
import { faUser } from '@fortawesome/free-solid-svg-icons';
import { CookieValueTypes } from 'cookies-next';
import { IQRCode } from 'next-qrcode/dist/useQRCode';
import React from 'react';

// Internal
import { ProfileForm, ProfileHeader, ProfileOrganisation } from '@/components/auth';
import { Block } from '@/components/ui/block-text';
import { FlexibleBox } from '@/components/ui/flexible-box';
import { Organisation, User } from '@/types';

export type ProfileProps = {
    renderUser: User | undefined;
    imagePreview: string | undefined;
    renderOrganisation: Organisation | undefined
    accessToken: CookieValueTypes
    handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
    handleImageUpload: (e: React.ChangeEvent<HTMLInputElement>) => void;
    handleSaveChanges: () => void;
    handleDeleteUser: () => void;
    Canvas: <T_1 extends HTMLCanvasElement>({ text, options, logo, }: IQRCode) => React.JSX.Element
}

export const Profile: React.FC<ProfileProps> = (props) => props.renderUser && (
    <Block className="page-content">
        <FlexibleBox
            title={`Profile Settings`}
            subtitle={`${props.renderUser.User_FirstName} ${props.renderUser.User_Surname}`}
            titleAction={
                <ProfileHeader />
            }
            icon={faUser}
            className="no-box w-auto inline-block"
            numberOfColumns={2}
        >
            <ProfileForm {...props} />
        </FlexibleBox>

        <ProfileOrganisation renderOrganisation={props.renderOrganisation} />
    </Block>
);
