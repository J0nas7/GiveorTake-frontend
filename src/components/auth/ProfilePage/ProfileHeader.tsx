import { Block, Text } from "@/components/ui/block-text";
import { faHouseChimney } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import Link from "next/link";
import React from "react";

export const ProfileHeader: React.FC = () => (
    <Block className="actions-wrapper">
        <Link
            href={`/`}
            className="blue-link action-button button-right"
        >
            <FontAwesomeIcon icon={faHouseChimney} />
            <Text variant="span">Go to Home</Text>
        </Link>
    </Block>
);
