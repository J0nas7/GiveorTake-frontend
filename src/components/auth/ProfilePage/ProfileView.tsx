"use client";

// External
import { useQRCode } from 'next-qrcode';
import React, { useEffect, useState } from 'react';

// Internal
import { Profile, ProfileProps } from '@/components/auth';
import { useUsersContext } from '@/contexts';
import { useCookies } from '@/hooks';
import { selectAuthUser, selectAuthUserOrganisation, selectAuthUserSeat, setSnackMessage, useAppDispatch, useTypedSelector } from '@/redux';
import { Organisation, User } from '@/types';

export const ProfileView: React.FC = () => {
    // Hooks
    const dispatch = useAppDispatch()
    const { saveUserChanges, removeUser } = useUsersContext()
    const authUserSeats = useTypedSelector(selectAuthUserSeat); // Redux
    const { Canvas } = useQRCode();
    const { getTheCookie } = useCookies()
    const accessToken = getTheCookie("accessToken")

    // State
    const authUserOrganisation = useTypedSelector(selectAuthUserOrganisation); // Redux
    const authUser = useTypedSelector(selectAuthUser); // Redux
    const [renderUser, setRenderUser] = useState<User | undefined>(undefined)
    const [renderOrganisation, setRenderOrganisation] = useState<Organisation | undefined>(undefined)
    const [imagePreview, setImagePreview] = useState<string | undefined>(renderUser?.User_ImageSrc);

    // Effects
    useEffect(() => {
        if (authUserSeats && authUserSeats.team?.projects?.[0]?.Project_ID) {
            setRenderOrganisation(authUserSeats.team?.organisation)
        } else if (authUserOrganisation && authUserOrganisation.teams?.[0].projects?.[0].Project_ID) {
            setRenderOrganisation(authUserOrganisation)
        }
    }, [authUserSeats, authUserOrganisation])

    // Set initial user data when the component mounts
    useEffect(() => {
        if (authUser) {
            setRenderUser(authUser);
            setImagePreview(authUser.User_ImageSrc);
        }
    }, [authUser]);

    // Handle input change
    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        if (renderUser) {
            const { name, value } = e.target;
            setRenderUser((prevState) => ({
                ...prevState!,
                [name]: value,
            }));
        }
    };

    // Handle image upload
    const handleImageUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
        const file = e.target.files?.[0];
        if (file) {
            const reader = new FileReader();
            reader.onloadend = () => {
                setImagePreview(reader.result as string);
                if (renderUser) {
                    setRenderUser((prevState) => ({
                        ...prevState!,
                        User_ImageSrc: reader.result as string,
                    }));
                }
            };
            reader.readAsDataURL(file);
        }
    };

    // Handle form submission (saving user changes)
    const handleSaveChanges = async () => {
        if (renderUser) {
            const saveChanges = await saveUserChanges(renderUser, 0)

            dispatch(setSnackMessage(
                saveChanges ? "User changes saved successfully!" : "Failed to save user changes."
            ))
        }
    };

    // Handle user deletion
    const handleDeleteUser = () => {
        if (renderUser && renderUser.User_ID) {
            removeUser(renderUser.User_ID, 0, "/");
        }
    }

    // Render
    const profileProps: ProfileProps = {
        renderUser,
        imagePreview,
        renderOrganisation,
        accessToken,
        handleChange,
        handleImageUpload,
        handleSaveChanges,
        handleDeleteUser,
        Canvas
    }

    return <Profile {...profileProps} />
};
