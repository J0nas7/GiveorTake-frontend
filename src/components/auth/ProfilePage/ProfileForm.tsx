import { Block } from "@/components/ui/block-text";
import styles from "@/core-ui/styles/modules/User.settings.module.scss";
import React from "react";
import { ProfileProps } from "./Profile";

type ProfileFormProps = Pick<
    ProfileProps,
    "renderUser" |
    "handleChange" |
    "handleImageUpload" |
    "handleSaveChanges" |
    "handleDeleteUser" |
    "imagePreview" |
    "accessToken" |
    "Canvas"
>

export const ProfileForm: React.FC<ProfileFormProps> = (props) => (
    <div className={styles.userDetailsForm}>
        <div className={styles.formGroup}>
            <label htmlFor="firstName" className={styles.label}>First Name</label>
            <input type="text" id="firstName" name="User_FirstName" value={props.renderUser?.User_FirstName} onChange={props.handleChange} className={styles.input} />
        </div>

        <div className={styles.formGroup}>
            <label htmlFor="surname" className={styles.label}>Surname</label>
            <input type="text" id="surname" name="User_Surname" value={props.renderUser?.User_Surname} onChange={props.handleChange} className={styles.input} />
        </div>

        <div className={styles.formGroup}>
            <label htmlFor="email" className={styles.label}>Email</label>
            <input type="email" id="email" name="User_Email" value={props.renderUser?.User_Email} onChange={props.handleChange} className={styles.input} />
        </div>

        <div className={styles.formGroup}>
            <label className={styles.label}>Profile Image</label>
            <div className={styles.profileImageContainer}>
                {props.imagePreview ? (
                    <img src={props.imagePreview} alt="User Profile" className={styles.userProfileImage} />
                ) : (
                    <div className={styles.noImage}>No image</div>
                )}
                <input type="file" onChange={props.handleImageUpload} className={styles.imageUploadInput} accept="image/*" />
            </div>
        </div>

        <div className={styles.formGroup}>
            <props.Canvas text={props.accessToken!.toString()} />
        </div>

        {props.renderUser?.User_DeletedAt ? (
            <div className={styles.deletedMessage}>
                <p>This user has been deleted.</p>
            </div>
        ) : (
            <Block className={styles.formActions}>
                <button onClick={props.handleSaveChanges} className="button-blue">Save Changes</button>
                <button onClick={props.handleDeleteUser} className="blue-link-light red-link-light">Delete User</button>
            </Block>
        )}
    </div>
);
