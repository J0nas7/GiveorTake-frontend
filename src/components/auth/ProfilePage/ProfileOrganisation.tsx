import { Heading } from "@/components/ui/heading";
import styles from "@/core-ui/styles/modules/User.settings.module.scss";
import Link from "next/link";
import React from "react";
import { ProfileProps } from "./Profile";

type ProfileOrganisationProps = Pick<
    ProfileProps,
    "renderOrganisation"
>

export const ProfileOrganisation: React.FC<ProfileOrganisationProps> = (props) => (
    <div className={styles.userTeamsContainer}>
        <Heading variant="h3" className={styles.teamsHeading}>Organisation this user is a part of</Heading>
        {props.renderOrganisation ? (
            <ul className={styles.teamsList}>
                <li className={styles.teamItem}>
                    <p>
                        <strong>Organisation Name:</strong>{" "}
                        <Link
                            href={`/organisation/${props.renderOrganisation.Organisation_ID}`}
                            className="blue-link-light"
                        >
                            {props.renderOrganisation.Organisation_Name}
                        </Link>
                    </p>
                </li>
            </ul>
        ) : (
            <p>This user is not part of any organisation.</p>
        )}
    </div>
);
