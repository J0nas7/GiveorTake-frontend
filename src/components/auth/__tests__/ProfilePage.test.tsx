import { TestProvider } from '@/__tests__/test-utils';
import '@testing-library/jest-dom';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Profile } from '../ProfilePage/Profile';
import { ProfileForm } from '../ProfilePage/ProfileForm';
import { ProfileHeader } from '../ProfilePage/ProfileHeader';
import { ProfileOrganisation } from '../ProfilePage/ProfileOrganisation';

void React.createElement

describe('ProfilePage Components', () => {
    describe('Profile', () => {
        const mockProps = {
            renderUser: {
                User_ID: 0,
                User_FirstName: 'John',
                User_Surname: 'Doe',
                User_Email: 'john.doe@example.com',
                User_Status: "1"
            },
            imagePreview: 'image-preview-url',
            renderOrganisation: {
                User_ID: 0,
                Organisation_ID: 1,
                Organisation_Name: 'Test Org'
            },
            accessToken: 'mockedAccessToken',
            handleChange: jest.fn(),
            handleImageUpload: jest.fn(),
            handleSaveChanges: jest.fn(),
            handleDeleteUser: jest.fn(),
            Canvas: () => <div>Canvas</div>,
        };

        const renderProfile = (props = {}) =>
            render(
                <TestProvider>
                    <Profile {...mockProps} {...props} />
                </TestProvider>
            );

        beforeEach(() => {
            jest.clearAllMocks();

            renderProfile()
        });

        it('renders profile title and user name', () => {
            expect(screen.getByText(/Profile Settings/i)).toBeInTheDocument();
            expect(screen.getByText('John Doe')).toBeInTheDocument();
        });

        it('renders image preview if available', () => {
            const image = screen.getByAltText(/User Profile/i);
            expect(image).toBeInTheDocument();
            expect(image).toHaveAttribute('src', 'image-preview-url');
        });

        it('renders organisation name', () => {
            expect(screen.getByText(/Test Org/i)).toBeInTheDocument();
        });

        it('calls handleImageUpload on file input change', () => {
            const fileInput = screen.getByLabelText(/Profile Image/i);
            const file = new File(['dummy'], 'avatar.png', { type: 'image/png' });
            fireEvent.change(fileInput, { target: { files: [file] } });
            expect(mockProps.handleImageUpload).toHaveBeenCalled();
        });

        it('renders null if renderUser is undefined', () => {
            cleanup(); // Clear the DOM from beforeEach render
            const { container } = renderProfile({ renderUser: undefined })

            expect(container.textContent?.trim()).toBe('');
        });
    });

    describe("ProfileForm", () => {
        const mockFormProps = {
            renderUser: {
                User_ID: 0,
                User_FirstName: "John",
                User_Surname: "Doe",
                User_Email: "john.doe@example.com",
                User_Status: "1",
            },
            imagePreview: "image-preview-url",
            handleChange: jest.fn(),
            handleImageUpload: jest.fn(),
            handleSaveChanges: jest.fn(),
            handleDeleteUser: jest.fn(),
            accessToken: "mockedAccessToken",
            Canvas: () => <div>Canvas</div>,
        };

        const renderProfileForm = (props = {}) =>
            render(
                <TestProvider>
                    <ProfileForm {...mockFormProps} {...props} />
                </TestProvider>
            );

        beforeEach(() => {
            jest.clearAllMocks();

            renderProfileForm()
        });

        it("renders form fields with correct initial values", () => {
            expect(screen.getByLabelText("First Name")).toBeInTheDocument();
            expect(screen.getByLabelText("First Name")).toHaveValue("John");

            expect(screen.getByLabelText("Surname")).toBeInTheDocument();
            expect(screen.getByLabelText("Surname")).toHaveValue("Doe");

            expect(screen.getByLabelText("Email")).toBeInTheDocument();
            expect(screen.getByLabelText("Email")).toHaveValue("john.doe@example.com");

            expect(screen.getByLabelText("Profile Image")).toBeInTheDocument();
        });

        it("calls handleChange when input fields change", () => {
            fireEvent.change(screen.getByLabelText("First Name"), {
                target: { value: "Jane" },
            });
            expect(mockFormProps.handleChange).toHaveBeenCalledTimes(1);

            fireEvent.change(screen.getByLabelText("Surname"), {
                target: { value: "Smith" },
            });
            expect(mockFormProps.handleChange).toHaveBeenCalledTimes(2);

            fireEvent.change(screen.getByLabelText("Email"), {
                target: { value: "jane.smith@example.com" },
            });
            expect(mockFormProps.handleChange).toHaveBeenCalledTimes(3);
        });

        it("calls handleImageUpload when profile image file input changes", () => {
            const file = new File(["dummy content"], "photo.png", { type: "image/png" });
            fireEvent.change(screen.getByLabelText("Profile Image"), {
                target: { files: [file] },
            });

            expect(mockFormProps.handleImageUpload).toHaveBeenCalledTimes(1);
        });

        it("displays image preview if imagePreview is provided", () => {
            const image = screen.getByAltText("User Profile") as HTMLImageElement;
            expect(image).toBeInTheDocument();
            expect(image.src).toContain("image-preview-url");
        });

        it("displays 'No image' if imagePreview is not provided", () => {
            renderProfileForm({ imagePreview: undefined })

            expect(screen.getByText("No image")).toBeInTheDocument();
        });

        it("renders the Canvas component", () => {
            expect(screen.getByText("Canvas")).toBeInTheDocument();
        });

        it("calls handleSaveChanges and handleDeleteUser on button clicks", () => {
            fireEvent.click(screen.getByText("Save Changes"));
            expect(mockFormProps.handleSaveChanges).toHaveBeenCalledTimes(1);

            fireEvent.click(screen.getByText("Delete User"));
            expect(mockFormProps.handleDeleteUser).toHaveBeenCalledTimes(1);
        });

        it("shows deleted message and hides buttons if user is deleted", () => {
            cleanup(); // Clear the DOM from beforeEach render
            renderProfileForm({ renderUser: { ...mockFormProps.renderUser, User_DeletedAt: "2023-01-01" } })

            expect(screen.getByText("This user has been deleted.")).toBeInTheDocument();
            expect(screen.queryByText("Save Changes")).toBeNull();
            expect(screen.queryByText("Delete User")).toBeNull();
        });
    });

    describe('ProfileHeader', () => {
        const renderProfileHeader = () =>
            render(
                <TestProvider>
                    <ProfileHeader />
                </TestProvider>
            );

        beforeEach(() => {
            jest.clearAllMocks();

            renderProfileHeader()
        });

        it('renders the go to home link with correct text', () => {
            const link = screen.getByRole('link', { name: /Go to Home/i });
            expect(link).toBeInTheDocument();
        });

        it('link points to the root URL "/"', () => {
            const link = screen.getByRole('link', { name: /Go to Home/i });
            expect(link).toHaveAttribute('href', '/');
        });

        it('renders the house icon', () => {
            // We can't directly select FontAwesomeIcon, but can check if SVG is in the link
            const link = screen.getByRole('link', { name: /Go to Home/i });
            const svg = link.querySelector('svg');
            expect(svg).toBeInTheDocument();
        });

        it('has expected CSS classes for styling', () => {
            const link = screen.getByRole('link', { name: /Go to Home/i });
            expect(link).toHaveClass('blue-link', 'action-button', 'button-right');
        });
    });

    describe("ProfileOrganisation", () => {
        const mockOrgProps = {
            renderOrganisation: {
                User_ID: 0,
                Organisation_ID: 1,
                Organisation_Name: "Test Org",
            },
        };

        const renderProfileOrganisation = (props = {}) =>
            render(
                <TestProvider>
                    <ProfileOrganisation {...mockOrgProps} {...props} />
                </TestProvider>
            );

        beforeEach(() => {
            jest.clearAllMocks();

            renderProfileOrganisation()
        });

        it("renders the section heading", () => {
            expect(
                screen.getByRole("heading", { name: /Organisation this user is a part of/i })
            ).toBeInTheDocument();
        });

        it("renders organisation name and link if organisation is provided", () => {
            const orgLink = screen.getByRole("link", { name: /Test Org/i });
            expect(orgLink).toBeInTheDocument();
            expect(orgLink).toHaveAttribute("href", "/organisation/1");

            expect(screen.getByText("Organisation Name:")).toBeInTheDocument();
        });

        it("renders message when organisation is not provided", () => {
            cleanup(); // Clear the DOM from beforeEach render
            renderProfileOrganisation({ renderOrganisation: undefined });

            expect(screen.getByText("This user is not part of any organisation.")).toBeInTheDocument();
        });

        it("does not render organisation list when organisation is not provided", () => {
            cleanup(); // Clear the DOM from beforeEach render
            renderProfileOrganisation({ renderOrganisation: undefined });

            expect(screen.queryByRole("list")).toBeNull();
            expect(screen.queryByRole("listitem")).toBeNull();
            expect(screen.queryByRole("link")).toBeNull();
        });

        it("matches snapshot when organisation is provided", () => {
            cleanup(); // Clear the DOM from beforeEach render
            const { container } = renderProfileOrganisation()

            expect(container).toMatchSnapshot();
        });

        it("matches snapshot when organisation is not provided", () => {
            cleanup(); // Clear the DOM from beforeEach render
            const { container } = renderProfileOrganisation({ renderOrganisation: undefined });

            expect(container).toMatchSnapshot();
        });
    });
});
