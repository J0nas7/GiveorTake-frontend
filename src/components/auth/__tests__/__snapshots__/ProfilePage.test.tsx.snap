// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage Components ProfileOrganisation matches snapshot when organisation is not provided 1`] = `
<div>
   
  <div
    class="userTeamsContainer"
  >
    <h3
      class=" teamsHeading"
    >
      Organisation this user is a part of
    </h3>
    <p>
      This user is not part of any organisation.
    </p>
  </div>
</div>
`;

exports[`ProfilePage Components ProfileOrganisation matches snapshot when organisation is provided 1`] = `
<div>
   
  <div
    class="userTeamsContainer"
  >
    <h3
      class=" teamsHeading"
    >
      Organisation this user is a part of
    </h3>
    <ul
      class="teamsList"
    >
      <li
        class="teamItem"
      >
        <p>
          <strong>
            Organisation Name:
          </strong>
           
          <a
            class="blue-link-light"
            href="/organisation/1"
          >
            Test Org
          </a>
        </p>
      </li>
    </ul>
  </div>
</div>
`;
